src/microbe_stage/CompoundCloudSystem.cs

```diff
@@ -213,134 +185,167 @@ public float AmountAvailable(Compound compound, Vector3 worldPosition, float fra
+            cloudInstance.Set("position", newPosition);
+            cloudInstance.Call("update_position", new Vector2(cloudGridCenter.X, cloudGridCenter.Z));
+        }
+    }
 
-            // No lock needed here now as AbsorbCompounds now uses atomic reads and updates
-            for (int x = (int)MathF.Round(cloudRelativeX - localGrabRadius); x <= xEnd; x += 1)
+    public int Resolution
+    {
+        get
+        {
+            if (cloudInstances.Count > 0)
             {
-                for (int y = (int)MathF.Round(cloudRelativeY - localGrabRadius); y <= yEnd; y += 1)
-                {
-                    // Negative coordinates are always outside the cloud area
-                    if (x < 0 || y < 0)
-                        continue;
+                return (int)cloudInstances[0].Get("resolution");
+            }
 
-                    // Circle check
-                    if (MathF.Pow(x - cloudRelativeX, 2) + MathF.Pow(y - cloudRelativeY, 2) >
-                        localGrabRadiusSquared)
-                    {
-                        // Not in it
-                        continue;
-                    }
+            return 4; // Default value
+        }
+    }
 
-                    // Then just need to check that it is within the cloud simulation array
-                    if (x < cloud.Size && y < cloud.Size)
-                    {
-                        // Absorb all compounds in the cloud
-                        cloud.AbsorbCompounds(x, y, storage, totals, delta, rate);
-                    }
-                }
+    public int Size
+    {
+        get
+        {
+            if (cloudInstances.Count > 0)
+            {
+                return (int)cloudInstances[0].Get("size");
             }
+
+            return 256; // Default value
         }
     }
 
-    public Vector3? FindCompoundNearPoint(Vector3 position, Compound compound, float searchRadius = 200,
-        float minConcentration = 120)
+    public Vector3? FindCompoundNearPoint(Vector3 position, Compound compound, float searchRadius = 200, float minConcentration = 120)
     {
         if (searchRadius < 1)
             throw new ArgumentException("searchRadius must be >= 1");
 
         int resolution = Resolution;
-
-        // This version is used when working with cloud local coordinates
         float localRadius = searchRadius / resolution;
-
+        float localRadiusSquared = localRadius * localRadius;
         float nearestDistanceSquared = float.MaxValue;
-
         Vector3? closestPoint = null;
+        int compoundID = GetCompoundID(compound);
 
-        foreach (var cloud in clouds)
+        foreach (var cloudInstance in cloudInstances)
         {
-            // Skip clouds that don't handle the target compound
-            if (!cloud.HandlesCompound(compound))
+            var compounds = (int[])cloudInstance.Call("get_compounds");
+
+            if (!Array.Exists(compounds, id => id == compoundID))
```

Comment:
This allocates a bit of memory, would be really nice if a core operation like this didn't need to constantly allocate memory.